brave/brave-ios
# Fix #7779: Support EIP-6963

## Summary of Changes
- Implements [EIP-6963](https://eips.ethereum.org/EIPS/eip-6963) for Ethereum DApps.
- ~Integrated [ethereum_provider.js](https://github.com/brave/brave-core/blob/master/components/brave_wallet/resources/ethereum_provider.js) from `brave-core` into [WalletEthereumProviderScript.js](https://github.com/brave/brave-ios/blob/development/Sources/Brave/Frontend/UserContent/UserScripts/Scripts_Dynamic/Scripts/Paged/WalletEthereumProviderScript.js) from `brave-ios` on `brave-ios`.~
    - EIP-6963 required `ethereum.on` to function, which was not available until we injected the `ethereum_provider.js` from BraveCore.
    - Previously the `ethereum_provider.js` from `brave-core` webpacked the `events` package into the script for us, now we have to webpack it ourselves so it's included in our `package.json` as a dependency & will be webpacked as part of our [bootstrap.sh](https://github.com/brave/brave-ios/blob/development/bootstrap.sh) script.~
- Updated `WalletEthereumProviderScript` so that the properties on the provider match `writable` behaviours of desktop except for `chainId`, `networkVersion`, and `selectedAddress` as we still need to [update these ourselves via javascript](https://github.com/brave/brave-ios/blob/development/Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController%2BWallet.swift#L465-L469).

Security/Privacy Review: https://github.com/brave/reviews/issues/1393

This pull request fixes #7779

## Submitter Checklist:


## Test Plan:
1. If compiling locally, you need to run bootstrap on this branch.
2. Verify that https://eip6963.org/ sees Brave Wallet with an icon and can connect.
3. Verify Ethereum DApp functionality still working as expected on multiple DApp sites.


## Screenshots:

https://github.com/brave/brave-ios/assets/5314553/61138900-1858-4f6b-b4f1-e4dd3d29e6f7


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
